packages/terminal/src/browser/terminal-widget-impl.ts

```diff
@@ -733,4 +732,9 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
     private disableEnterWhenAttachCloseListener(): boolean {
         return this.isAttachedCloseListener;
     }
+
+    // define exit status before firing onTermDidClose
+    protected setExitStatus(code?: number): void {
+        this._exitStatus = { code };
+    }
```

Comment:
```suggestion
    protected fireDidTermClose(exitCode?: number): void {
        this._exitStatus = { code: exitCode };
        this.onDidTermClose.fire(this);
    }
```

Reply:
I believe the original code had flaws, making `TerminalWidgetImpl#onTermDidClose` fire twice, once in `TerminalWatcher#onTerminalExit` callback, and once in `TerminalWidget#dispose` method that is called in many places, including this same callback. That caused the plugin side to receive  the `$terminalClosed` message twice, which resulted in always having `TerminalExitStatus#code` undefined, even though the terminal exited normally (code 0).

My fix was to keep the firing of `onTermDidClose` in the `dispose` method, though now that I'm thinking about it, there actually might be cases where `_exitStatus` is not set for some reason. I think the best way to fix all of these issues would be to check if `_exitStatus` was defined (and give it a default value otherwise) before firing `onTermDidClose` in the `dispose` method.

Reply:
Resolved in commit : [Fixed exitStatus undefined when closing terminal manually](https://github.com/eclipse-theia/theia/pull/11175/commits/962e252d55d4608f218968bd2be5e7dfcae304f4)